File: PetMedicalCare/BackEnd/SoftPetBusiness/src/main/java/pe/edu/pucp/softpet/bo/MascotaBo.java
package pe.edu.pucp.softpet.bo;

import java.sql.Date;
import java.util.ArrayList;
import pe.edu.pucp.softpet.daoImp.MascotaDaoImpl;
import pe.edu.pucp.softpet.dto.mascotas.MascotaDto;
import pe.edu.pucp.softpet.dto.personas.PersonaDto;

public class MascotaBo {

    private final MascotaDaoImpl mascotaDao;

    public MascotaBo() {
        this.mascotaDao = new MascotaDaoImpl();
    }

    public Integer insertar(int personaId, String nombre, String especie,
            String sexo, String raza, String color, String fechaDefuncion,
            boolean activo) {

        MascotaDto mascota = new MascotaDto();

        PersonaDto persona = new PersonaDto();
        persona.setPersonaId(personaId);

        mascota.setPersona(persona);
        mascota.setNombre(nombre);
        mascota.setEspecie(especie);
        mascota.setSexo(sexo);
        mascota.setRaza(raza);
        mascota.setColor(color);
        if (fechaDefuncion != null) {
            mascota.setFechaDefuncion(Date.valueOf(fechaDefuncion));
        }
        mascota.setActivo(activo);

        return this.mascotaDao.insertar(mascota);
    }

    public Integer modificar(int mascotaId, int personaId, String nombre, String especie,
            String sexo, String raza, String color, String fechaDefuncion,
            boolean activo) {

        MascotaDto mascota = new MascotaDto();

        PersonaDto persona = new PersonaDto();
        persona.setPersonaId(personaId);

        mascota.setMascotaId(mascotaId);
        mascota.setPersona(persona);
        mascota.setNombre(nombre);
        mascota.setEspecie(especie);
        mascota.setSexo(sexo);
        mascota.setRaza(raza);
        mascota.setColor(color);
        if (fechaDefuncion != null) {
            mascota.setFechaDefuncion(Date.valueOf(fechaDefuncion));
        }
        mascota.setActivo(activo);

        return this.mascotaDao.modificar(mascota);
    }

    public Integer eliminar(int mascotaId) {
        MascotaDto mascota = new MascotaDto();
        mascota.setMascotaId(mascotaId);
        return this.mascotaDao.eliminar(mascota);
    }

    public MascotaDto obtenerPorId(int mascotaId) {
        return this.mascotaDao.obtenerPorId(mascotaId);
    }

    public ArrayList<MascotaDto> listarTodos() {
        return this.mascotaDao.listarTodos();
    }

    public ArrayList<MascotaDto> ListasBusquedaAvanzada(
            String nombreMascota, String especie, String nombreDeLaPersona, Integer estadoActivo) {
        return (ArrayList<MascotaDto>) mascotaDao.ListasBusquedaAvanzada(
                nombreMascota == null ? "" : nombreMascota,
                especie == null ? "" : especie,
                nombreDeLaPersona == null ? "" : nombreDeLaPersona,
                estadoActivo);
    }

    public ArrayList<MascotaDto> listarPorIdPersona(int personaId) {
        return this.mascotaDao.listarPorIdPersona(personaId);
    }

    public ArrayList<MascotaDto> listarMascotasActivas() {
        return this.mascotaDao.listarMascotasActivas();
    }

    public int VerificarSiLaMascotaTieneInformacion(int idServicio) {
        return this.mascotaDao.VerificarSiLaMascotaTieneInformacion(idServicio);
    }
}

File: Mis últimos pasos/SoftPetWA/SoftPetBussiness/SOAPBo/GestionNOLogistica/MascotaBo.cs

using SoftPetBussiness.MascotaClient;
using System;
using System.Collections.Generic;
using System.Linq;


namespace SoftPetBusiness
{
    public class MascotaBO
    {
        private MascotasClient clienteSOAP;

        public MascotaBO()
        {
            this.clienteSOAP = new MascotasClient();
        }

        // Insertar mascota (parámetros sueltos)
        public int Insertar(int personaId, string nombre, string especie, string sexo,
                            string raza, string color, string fechaDefuncion, bool activo)
        {
            // CORRECCIÓN: Si viene vacío, enviamos null al servicio
            if (string.IsNullOrEmpty(fechaDefuncion))
            {
                fechaDefuncion = null;
            }

            return this.clienteSOAP.insertar_mascota(
                personaId, nombre, especie, sexo, raza, color, fechaDefuncion, activo
            );
        }

        // Modificar mascota
        public int Modificar(int mascotaId, int personaId, string nombre, string especie, string sexo,
                             string raza, string color, string fechaDefuncion, bool activo)
        {
            // CORRECCIÓN: Si viene vacío, enviamos null al servicio
            if (string.IsNullOrEmpty(fechaDefuncion))
            {
                fechaDefuncion = null;
            }

            return this.clienteSOAP.modificar_mascota(
                mascotaId, personaId, nombre, especie, sexo, raza, color, fechaDefuncion, activo
            );
        }

        // Eliminar mascota
        public int Eliminar(int mascotaId)
        {
            return this.clienteSOAP.eliminar_mascota(mascotaId);
        }

        // Obtener mascota por ID
        public mascotaDto ObtenerPorId(int mascotaId)
        {
            return this.clienteSOAP.obtenerPorId_mascota(mascotaId);
        }

        // Listar todas las mascotas
        public List<mascotaDto> ListarTodos()
        {
            return this.clienteSOAP.listar_mascotas().ToList<mascotaDto>();
        }

        // Insertar mascota usando DTO
        //public int Insertar(mascotaDto mascota)
        //{
        //    return this.clienteSOAP.insertar_mascota(
        //        mascota.persona.personaId,
        //        mascota.nombre,
        //        mascota.especie,
        //        mascota.sexo,
        //        mascota.raza,
        //        mascota.color,
        //        // Si en el proxy fechaDefuncion es DateTime?, usar mascota.fechaDefuncion?.ToString("yyyy-MM-dd")
        //        mascota.fechaDefuncion.ToString(),
        //        mascota.activo
        //    );
        //}
        public int Insertar(mascotaDto mascota, DateTime fechadefuncion)
        {
            string fechaDefuncionStr;
            if (fechadefuncion == null)
            {
                fechaDefuncionStr = DateTime.MinValue.ToString("yyyy-MM-dd");
            }
            else
            {
                fechaDefuncionStr = fechadefuncion.ToString("yyyy-MM-dd");
            }


            return this.clienteSOAP.insertar_mascota(
                mascota.persona.personaId,
                mascota.nombre,
                mascota.especie,
                mascota.sexo,
                mascota.raza,
                mascota.color,
                // Si en el proxy fechaDefuncion es DateTime?, usar mascota.fechaDefuncion?.ToString("yyyy-MM-dd")
                fechaDefuncionStr,
                mascota.activo
            );
        }
        //public int Insertar(mascotaDto mascota, string fechadefuncion)
        //{
        //    string fechaDefuncionStr;
        //    if (fechadefuncion == null || fechadefuncion == "")
        //    {
        //        fechaDefuncionStr = DateTime.MinValue.ToString("yyyy-MM-dd");
        //    }
        //    fechaDefuncionStr=fechadefuncion;
        //    return this.clienteSOAP.insertar_mascota(
        //        mascota.persona.personaId,
        //        mascota.nombre,
        //        mascota.especie,
        //        mascota.sexo,
        //        mascota.raza,
        //        mascota.color,
        //        // Si en el proxy fechaDefuncion es DateTime?, usar mascota.fechaDefuncion?.ToString("yyyy-MM-dd")
        //        fechaDefuncionStr,
        //        mascota.activo
        //    );
        //}

        // Modificar mascota usando DTO
        public int Modificar(mascotaDto mascota)
        {
            return this.clienteSOAP.modificar_mascota(
                mascota.mascotaId,
                mascota.persona.personaId,
                mascota.nombre,
                mascota.especie,
                mascota.sexo,
                mascota.raza,
                mascota.color,
                mascota.fechaDefuncion.ToString(),
                mascota.activo
            );
        }
        public int Modificar(mascotaDto mascota, DateTime fechadefuncion)
        {
            string fechaDefuncionStr;
            if (fechadefuncion == null)
            {
                fechaDefuncionStr = DateTime.MinValue.ToString("yyyy-MM-dd");
            }
            else
            {
                fechaDefuncionStr = fechadefuncion.ToString("yyyy-MM-dd");
            }
            return this.clienteSOAP.modificar_mascota(
                mascota.mascotaId,
                mascota.persona.personaId,
                mascota.nombre,
                mascota.especie,
                mascota.sexo,
                mascota.raza,
                mascota.color,
                fechaDefuncionStr
                ,
                mascota.activo
            );
        }
        public List<mascotaDto> ListarPorIdPersona(int personaId)
        {
            var resultado = this.clienteSOAP.listar_mascotas_por_persona(personaId);
            return (resultado ?? new mascotaDto[0]).ToList<mascotaDto>();
        }
        public List<mascotaDto> ListarActivos()
        {
            return this.clienteSOAP.listar_mascotas_activas().ToList<mascotaDto>();
        }

        public List<mascotaDto> ListarBusquedaAvanzada(
            string nombreMascota, string especie, string nombrePersona, int estadoActivo)
        {
            var resultado = this.clienteSOAP.ListasBusquedaAvanzada(nombreMascota, especie, nombrePersona, estadoActivo);
            if (resultado == null) return new List<mascotaDto>();
            return resultado.ToList<mascotaDto>();
        }
        public List<mascotaDto> listartodos(
            mascotaDto mascota)
        {
            return this.clienteSOAP.listar_mascotas().ToList<mascotaDto>();
        }
    }
}
